### PR TITLE
Experimental end-to-end encryption support

### PR DESCRIPTION
**Pull request by lukasIO**

### CODE REVIEW

**Comment by initpwn** [comment]
Status of this PR?

File: src/e2ee/worker/e2ee.worker.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  }
+}
+
+function setupCryptorErrorEvents(cryptor: Cryptor) {
+  cryptor.on('cryptorError', (error) => {
+    const msg: ErrorMessage = {
+      kind: 'error',
+      data: { error },
+    };
+    postMessage(msg);
```
Comment (max-b) [inline src/e2ee/worker/e2ee.worker.ts:188]: I believe this is an instance of attempting to pass an error w/ postMessage which browsers generally don't allow?
https://developer.mozilla.org/en-US/docs/Web/API/Worker/postMessage
https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm

File: src/e2ee/worker/Cryptor.ts [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+      .pipeThrough(transformStream)
+      .pipeTo(writable)
+      .catch((e) => {
+        const errorMsg: ErrorMessage = {
+          kind: 'error',
+          data: {
+            error: new E2EEError(e.message, E2EEErrorReason.InternalError),
+          },
+        };
+        postMessage(errorMsg);
```
Comment (max-b) [inline src/e2ee/worker/Cryptor.ts:148]: Same issue as https://github.com/livekit/client-sdk-js/pull/557/files#r1156612686 I believe

File: src/e2ee/worker/e2ee.worker.ts (hunk 1 above)
Comment (lukasIO) [reply src/e2ee/worker/e2ee.worker.ts]: thanks for the reminder 🙏

File: yarn.lock [hunk 3]
```diff
@@ -4466,6 +4454,12 @@ glob@^7.0.3, glob@^7.1.3, glob@^7.1.4:
     once "^1.3.0"
     path-is-absolute "^1.0.0"
 
+<<<<<<< HEAD
```
Comment (cloudwebrtc) [inline yarn.lock:4457]: There are some merge conflicts in the .lock file, you may need to override yarn.lock

File: src/room/participant/Participant.ts [hunk 4]
```diff
@@ -68,6 +68,10 @@ export default class Participant extends (EventEmitter as new () => TypedEmitter
 
   private _connectionQuality: ConnectionQuality = ConnectionQuality.Unknown;
 
+  get isEncrypted() {
+    return this.tracks.size > 0 && Array.from(this.tracks.values()).every((tr) => tr.isEncrypted);
```
Comment (davidzhao) [inline src/room/participant/Participant.ts:72]: so participant.isEncrypted is only true when every track is encrypted?

Comment (lukasIO) [reply src/room/participant/Participant.ts:72]: yeah, that seems the like the least confusing option. In theory all tracks published by a participant should be encrypted anyways after enabling e2ee. 
Anything that you would think is unexpected with that approach?

Comment (davidzhao) [reply src/room/participant/Participant.ts:72]: Nope.. just thinking through how this option would be used. But that makes sense to me.

**Review by cloudwebrtc — APPROVED** [review]
lgtm

**Comment by lukasIO** [comment]
confirmed this is working in conjunction with track-processors. Will cut a new patch release for livekit-client first before merging this in as a minor change

**Pull request merged**